pkg/dynamicdata/dynamicdata.go

```diff
@@ -0,0 +1,362 @@
+	if err != nil {
+		return nil, err
+	}
+
+	sync := &Sync{
+		opaConfig:          bs,
+		kubeconfig:         kubeconfig,
+		opaAuth:            opaAuth,
+		opaURL:             opaURL,
+		ignoreNs:           ignoreNs,
+		analysisEntrypoint: analysisEntrypoint,
+		replicatePath:      replicatePath,
+		logger:             logger,
+		running:            make(map[types.ResourceType]*cancellableSync),
+	}
+
+	return sync, nil
+}
+
+func (s *Sync) Run(ctx context.Context) error {
+
+	s.logger.Debug("Loading kubeconfig for API server")
+	client, err := dynamic.NewForConfig(s.kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debug("Resolving resource names to resource types")
+	rts, err := resolveResourceTypes(s.kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	s.logger.Debug("Starting analyzer")
+	analyzer, err := newAnalyzer(ctx, s.opaConfig, s.replicatePath, s.analysisEntrypoint, s.logger)
+	if err != nil {
+		return err
+	}
+
+	go s.loop(ctx, analyzer, rts, client)
+
+	return nil
+}
+
+func (s *Sync) Ready() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.ready {
+		s.logger.Debug("Sync is not ready")
+		return false
+	}
+	for rt, r := range s.running {
+		if !r.sync.Ready() {
+			s.logger.Debug("Replicator for %v is not ready", rt)
+			return false
+		}
+	}
+	return true
+}
+
+func (s *Sync) loop(ctx context.Context, a *analyzer, rts map[string]types.ResourceType, client *dynamic.DynamicClient) {
+	for {
+		s.logger.Debug("Sync waiting for analysis result")
+		select {
+		case result := <-a.C:
+			s.logger.Debug("Sync processing analysis result: %v", result)
+			s.processAnalysisResult(ctx, result, rts, client)
+		case <-ctx.Done():
+			s.logger.Debug("Sync shutting down")
+		}
+	}
+}
+
+func (s *Sync) processAnalysisResult(ctx context.Context, result analysisResult, rts map[string]types.ResourceType, client *dynamic.DynamicClient) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If any of the refs cannot be mapped to gvk then give up.
+	for _, ref := range result.Refs {
+		if _, ok := rts[ref.Resource]; !ok {
+			logrus.Errorf("Cannot resolve Kubernetes resource %q to group/version/resource for dynamic data replication", ref.Resource)
+			s.ready = false
+			return
+		}
+	}
+
+	// Otherwise, create and delete data syncs accordingly.
+	s.ready = true
+	create := map[types.ResourceType]struct{}{}
+
+	for _, ref := range result.Refs {
+		rt := rts[ref.Resource]
+		create[rt] = struct{}{}
+		if _, ok := s.running[rt]; !ok {
+			s.logger.Debug("Starting data replication for %v", rt)
+			sync := data.NewFromInterface(client, opa.New(s.opaURL, s.opaAuth).Prefix(s.replicatePath), rt, data.WithIgnoreNamespaces(s.ignoreNs))
+			ctx, cancel := context.WithCancel(ctx)
+			s.running[rt] = &cancellableSync{cancel: cancel, sync: sync}
+			go sync.RunContext(ctx)
+		} else {
+			s.logger.Debug("Data replication for %v already started", rt)
+		}
+	}
+
+	for rt, sync := range s.running {
+		if _, ok := create[rt]; !ok {
```

Comment:
Maybe I'm missing something but how can keys in `s.running` and `create` differ?

Reply:
Torin helped to clarify this out-of-band. Thanks!